File: 46.permutations.py
# https://leetcode.com/problems/permutations/
# @leet start
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []

        def backtrack(curr: list[int], visited: list[bool]):
            if len(curr) == n:
                ans.append(curr.copy())
            else:
                for i in range(n):
                    if not visited[i]:
                        curr.append(nums[i])
                        visited[i] = True
                        backtrack(curr, visited)
                        curr.pop()
                        visited[i] = False

        backtrack([], [False] * n)
        return ans


# @leet end

File: 46.permutations.go
package leetcode

// @leet start
func permute(nums []int) [][]int {
	n := len(nums)
	ans := [][]int{}
	visited := make([]bool, n)
	curr := []int{}
	backtrack(&nums, &ans, &curr, &visited)

	return ans
}

func backtrack(nums *[]int, ans *[][]int, curr *[]int, visited *[]bool) {
	if len(*curr) == len(*visited) {
		currCopy := make([]int, len(*curr))
		copy(currCopy, *curr)
		*ans = append(*ans, currCopy)
		return
	}

	for i := range len(*nums) {
		if !(*visited)[i] {
			*curr = append(*curr, (*nums)[i])
			(*visited)[i] = true
			backtrack(nums, ans, curr, visited)
			(*visited)[i] = false
			*curr = (*curr)[:len(*curr)-1]
		}

	}

}

// @leet end

